File: apps/backend/ModuAgent/evolution/registry/versioned_store.py
from __future__ import annotations

import importlib
import inspect
import json
import logging
import os
from typing import Any, Dict, List, Optional

logger = logging.getLogger(__name__)


class VersionedComponentStore:
    """组件版本快照存储。"""

    def __init__(self, storage_path: str = "evolution/versions"):
        self._storage_path = storage_path

    def _get_component_dir(self, component_name: str) -> str:
        """获取组件的存储目录路径。"""
        return os.path.join(self._storage_path, component_name)

    def _get_version_file_path(self, component_name: str, version: str) -> str:
        """获取指定版本的JSON文件路径。"""
        return os.path.join(self._get_component_dir(component_name), f"{version}.json")

    def _get_versions_index_path(self, component_name: str) -> str:
        """获取版本索引文件路径。"""
        return os.path.join(self._get_component_dir(component_name), "_versions.json")

    def _load_versions_index(self, component_name: str) -> List[str]:
        """加载版本索引列表。"""
        index_path = self._get_versions_index_path(component_name)
        if os.path.exists(index_path):
            with open(index_path, "r", encoding="utf-8") as f:
                return json.load(f)
        return []

    def _save_versions_index(self, component_name: str, versions: List[str]) -> None:
        """保存版本索引列表。"""
        component_dir = self._get_component_dir(component_name)
        os.makedirs(component_dir, exist_ok=True)
        index_path = self._get_versions_index_path(component_name)
        with open(index_path, "w", encoding="utf-8") as f:
            json.dump(versions, f, ensure_ascii=False, indent=2)

    @staticmethod
    def _serialize_component_config(component: Any) -> Optional[Dict[str, Any]]:
        """序列化组件配置（提取构造参数）。

        通过 inspect 获取 __init__ 参数签名，从对象属性中匹配参数值。
        存储类的模块路径和参数配置，回滚时通过反射重建。

        Args:
            component: 组件实例

        Returns:
            包含 module_path, class_name, init_params 的字典，
            无法序列化时返回 None
        """
        if component is None:
            return None

        try:
            cls = type(component)
            module_path = cls.__module__
            class_name = cls.__qualname__

            init_params: Dict[str, Any] = {}
            try:
                sig = inspect.signature(cls.__init__)
                for param_name in sig.parameters:
                    if param_name == "self":
                        continue
                    attr_name = f"_{param_name}"
                    if hasattr(component, attr_name):
                        value = getattr(component, attr_name)
                        if _is_json_serializable(value):
                            init_params[param_name] = value
                    elif hasattr(component, param_name):
                        value = getattr(component, param_name)
                        if _is_json_serializable(value):
                            init_params[param_name] = value
            except (ValueError, TypeError):
                pass

            if not init_params:
                init_params = {}
                for attr_name in dir(component):
                    if attr_name.startswith("_") or attr_name.startswith("__"):
                        continue
                    try:
                        value = getattr(component, attr_name)
                        if callable(value):
                            continue
                        if _is_json_serializable(value):
                            init_params[attr_name] = value
                    except Exception:
                        continue

            return {
                "module_path": module_path,
                "class_name": class_name,
                "init_params": init_params,
            }
        except Exception as e:
            logger.warning("Failed to serialize component config: %s", str(e))
            return None

    @staticmethod
    def _deserialize_component_config(component_config: Dict[str, Any]) -> Optional[Any]:
        """从配置反序列化重建组件实例。

        Args:
            component_config: _serialize_component_config 产出的配置字典

        Returns:
            重建的组件实例，失败时返回 None
        """
        if not component_config:
            return None

        try:
            module_path = component_config.get("module_path")
            class_name = component_config.get("class_name")
            init_params = component_config.get("init_params", {})

            if not module_path or not class_name:
                logger.warning("Invalid component config: missing module/class info")
                return None

            module = importlib.import_module(module_path)
            cls = module
            for name_part in class_name.split("."):
                cls = getattr(cls, name_part)

            try:
                sig = inspect.signature(cls.__init__)
                valid_params = {}
                for param_name in sig.parameters:
                    if param_name == "self":
                        continue
                    if param_name in init_params:
                        valid_params[param_name] = init_params[param_name]
                instance = cls(**valid_params)
            except (ValueError, TypeError):
                instance = cls(**init_params)

            logger.info(
                "Reconstructed component: %s.%s with %d params",
                module_path, class_name, len(init_params),
            )
            return instance
        except Exception as e:
            logger.warning("Failed to deserialize component: %s", str(e))
            return None

    def save_version(
        self,
        component_name: str,
        version: str,
        state: Dict[str, Any],
        metadata: Dict[str, Any],
        category: str = "",
        component: Any = None,
    ) -> None:
        """保存组件版本快照。

        Args:
            component_name: 组件名称
            version: 版本号
            state: 组件状态配置
            metadata: 元数据
            category: 组件分类（用于回滚时调用 registry.swap_component）
            component: 组件实例（用于回滚时恢复）
        """
        component_dir = self._get_component_dir(component_name)
        os.makedirs(component_dir, exist_ok=True)

        component_config = self._serialize_component_config(component)

        version_file_path = self._get_version_file_path(component_name, version)
        version_data = {
            "version": version,
            "state": state,
            "metadata": metadata,
            "category": category,
            "component_config": component_config,
        }
        with open(version_file_path, "w", encoding="utf-8") as f:
            json.dump(version_data, f, ensure_ascii=False, indent=2, default=str)

        versions = self._load_versions_index(component_name)
        if version not in versions:
            versions.append(version)
            self._save_versions_index(component_name, versions)

        logger.info("Saved version %s for component %s", version, component_name)

    def get_version(
        self,
        component_name: str,
        version: str,
    ) -> Optional[Dict[str, Any]]:
        """获取指定版本快照。

        从 component_config 反序列化重建 component 对象，
        保持与旧版 API 的兼容性（返回 dict 中包含 component 字段）。
        """
        version_file_path = self._get_version_file_path(component_name, version)
        if not os.path.exists(version_file_path):
            logger.warning("Version %s not found for component %s", version, component_name)
            return None

        with open(version_file_path, "r", encoding="utf-8") as f:
            version_data = json.load(f)

        component_config = version_data.get("component_config")
        if component_config and "component" not in version_data:
            component = self._deserialize_component_config(component_config)
            version_data["component"] = component

        return version_data

    def list_versions(
        self,
        component_name: str,
    ) -> List[str]:
        """列出组件的所有版本。"""
        return self._load_versions_index(component_name)

    def get_latest_version(
        self,
        component_name: str,
    ) -> Optional[str]:
        """获取组件的最新版本号。"""
        versions = self._load_versions_index(component_name)
        if not versions:
            return None
        return versions[-1]


def _is_json_serializable(value: Any) -> bool:
    """检查值是否可以被 JSON 序列化。

    Args:
        value: 待检查的值

    Returns:
        True 如果值可以被 JSON 序列化，否则 False
    """
    if value is None:
        return True
    if isinstance(value, (bool, int, float, str)):
        return True
    if isinstance(value, (list, tuple)):
        return all(_is_json_serializable(v) for v in value)
    if isinstance(value, dict):
        return all(
            isinstance(k, str) and _is_json_serializable(v)
            for k, v in value.items()
        )
    return False

File: packages/modu-agent/src/evolution/versioned-store.ts
// 对应 Python: evolution/registry/versioned_store.py
// VersionedComponentStore: 组件版本快照存储

import fs from 'fs'
import path from 'path'

const logger = {
  info: (msg: string, ...args: any[]) => console.info(`[versioned-store] ${msg}`, ...args),
  warning: (msg: string, ...args: any[]) => console.warn(`[versioned-store] ${msg}`, ...args),
  error: (msg: string, ...args: any[]) => console.error(`[versioned-store] ${msg}`, ...args),
  debug: (msg: string, ...args: any[]) => console.debug(`[versioned-store] ${msg}`, ...args),
}

/**
 * 检查值是否可以被 JSON 序列化。
 *
 * 对应 Python _is_json_serializable。
 */
function isJsonSerializable(value: any): boolean {
  if (value === null || value === undefined) {
    return true
  }
  if (typeof value === 'boolean' || typeof value === 'number' || typeof value === 'string') {
    return true
  }
  if (Array.isArray(value)) {
    return value.every(isJsonSerializable)
  }
  if (typeof value === 'object') {
    return Object.entries(value).every(
      ([k, v]) => typeof k === 'string' && isJsonSerializable(v),
    )
  }
  return false
}

/**
 * 组件版本快照存储。
 *
 * 对应 Python VersionedComponentStore。
 *
 * 适配说明：Python 版通过 inspect.signature 反射构造函数参数签名，
 * 并通过 importlib.import_module 动态导入模块重建组件实例。
 * TS/ESM 不支持运行时参数签名反射，也不支持同步动态 import()，
 * 因此：
 *   - 序列化：扫描对象自身可枚举属性（_paramName / paramName 模式）
 *   - 反序列化：优先从内存缓存取回组件实例；磁盘反序列化返回 null
 *   - 内存缓存保证同进程内 rollback 可用（与 Python 同进程行为等价）
 */
export class VersionedComponentStore {
  private _storagePath: string
  // 内存缓存：key = "componentName:version" → component 实例
  // ESM 无同步动态 import，缓存保证同进程内回滚可用
  private _componentCache: Map<string, any> = new Map()

  constructor(storagePath: string = 'evolution/versions') {
    this._storagePath = storagePath
  }

  /** 获取组件的存储目录路径。 */
  private _getComponentDir(componentName: string): string {
    return path.join(this._storagePath, componentName)
  }

  /** 获取指定版本的 JSON 文件路径。 */
  private _getVersionFilePath(componentName: string, version: string): string {
    return path.join(this._getComponentDir(componentName), `${version}.json`)
  }

  /** 获取版本索引文件路径。 */
  private _getVersionsIndexPath(componentName: string): string {
    return path.join(this._getComponentDir(componentName), '_versions.json')
  }

  /** 加载版本索引列表。 */
  private _loadVersionsIndex(componentName: string): string[] {
    const indexPath = this._getVersionsIndexPath(componentName)
    if (fs.existsSync(indexPath)) {
      try {
        return JSON.parse(fs.readFileSync(indexPath, 'utf-8'))
      } catch (e) {
        logger.warning('Failed to parse versions index for %s: %s', componentName, String(e))
        return []
      }
    }
    return []
  }

  /** 保存版本索引列表。 */
  private _saveVersionsIndex(componentName: string, versions: string[]): void {
    const componentDir = this._getComponentDir(componentName)
    fs.mkdirSync(componentDir, { recursive: true })
    const indexPath = this._getVersionsIndexPath(componentName)
    fs.writeFileSync(indexPath, JSON.stringify(versions, null, 2), 'utf-8')
  }

  /**
   * 序列化组件配置（提取构造参数）。
   *
   * 适配 TS：无法通过 inspect.signature 获取构造函数参数签名，
   * 改为扫描对象自身可枚举属性，匹配 _paramName / paramName 模式。
   * 存储 class_name 和 init_params，回滚时通过内存缓存或反射重建。
   */
  static _serializeComponentConfig(component: any): Record<string, any> | null {
    if (component === null || component === undefined) {
      return null
    }

    try {
      const cls = component.constructor
      const className = cls?.name ?? 'Unknown'
      // TS 没有 __module__，module_path 留空
      const modulePath: string | undefined = undefined

      const initParams: Record<string, any> = {}

      // 扫描对象自身属性（对应 Python 的 dir() fallback 分支）
      const ownProps = Object.getOwnPropertyNames(component)
      for (const attrName of ownProps) {
        // 跳过内部/双下划线属性
        if (attrName.startsWith('__')) continue

        try {
          const value = component[attrName]
          if (typeof value === 'function') continue
          if (isJsonSerializable(value)) {
            // 去除前导下划线以匹配构造参数名（Python _paramName → paramName）
            const cleanName = attrName.startsWith('_') ? attrName.slice(1) : attrName
            initParams[cleanName] = value
          }
        } catch {
          continue
        }
      }

      return {
        module_path: modulePath,
        class_name: className,
        init_params: initParams,
      }
    } catch (e) {
      logger.warning('Failed to serialize component config: %s', String(e))
      return null
    }
  }

  /**
   * 从配置反序列化重建组件实例。
   *
   * 适配 TS：ESM 不支持同步 import()，无法像 Python importlib 那样
   * 在同步方法中动态加载模块。此方法返回 null，实际组件实例
   * 通过内存缓存（_componentCache）在 get_version 中取回。
   */
  static _deserializeComponentConfig(componentConfig: Record<string, any>): any {
    if (!componentConfig) {
      return null
    }

    const modulePath = componentConfig.module_path
    const className = componentConfig.class_name
    const _initParams = componentConfig.init_params ?? {}

    if (!modulePath || !className) {
      logger.warning('Invalid component config: missing module/class info')
      return null
    }

    // ESM 无同步动态 import，无法在此重建实例。
    // 组件实例由 _componentCache 在 get_version 中提供。
    logger.warning(
      'Cannot deserialize component synchronously (ESM limitation): %s.%s',
      modulePath,
      className,
    )
    return null
  }

  /**
   * 保存组件版本快照。
   *
   * @param componentName 组件名称
   * @param version 版本号
   * @param state 组件状态配置
   * @param metadata 元数据
   * @param category 组件分类（用于回滚时调用 registry.swapComponent）
   * @param component 组件实例（用于回滚时恢复）
   */
  saveVersion(
    componentName: string,
    version: string,
    state: Record<string, any>,
    metadata: Record<string, any>,
    category: string = '',
    component: any = null,
  ): void {
    const componentDir = this._getComponentDir(componentName)
    fs.mkdirSync(componentDir, { recursive: true })

    const componentConfig = VersionedComponentStore._serializeComponentConfig(component)

    const versionFilePath = this._getVersionFilePath(componentName, version)
    const versionData: Record<string, any> = {
      version,
      state,
      metadata,
      category,
      component_config: componentConfig,
    }
    fs.writeFileSync(
      versionFilePath,
      JSON.stringify(versionData, null, 2),
      'utf-8',
    )

    // 内存缓存组件实例（ESM 适配：同进程内 rollback 可用）
    if (component !== null && component !== undefined) {
      this._componentCache.set(`${componentName}:${version}`, component)
    }

    const versions = this._loadVersionsIndex(componentName)
    if (!versions.includes(version)) {
      versions.push(version)
      this._saveVersionsIndex(componentName, versions)
    }

    logger.info('Saved version %s for component %s', version, componentName)
  }

  /**
   * 获取指定版本快照。
   *
   * 优先从内存缓存取回 component 实例；
   * 若缓存未命中，尝试从 component_config 反序列化（ESM 下通常返回 null）。
   */
  getVersion(
    componentName: string,
    version: string,
  ): Record<string, any> | null {
    const versionFilePath = this._getVersionFilePath(componentName, version)
    if (!fs.existsSync(versionFilePath)) {
      logger.warning('Version %s not found for component %s', version, componentName)
      return null
    }

    let versionData: Record<string, any>
    try {
      versionData = JSON.parse(fs.readFileSync(versionFilePath, 'utf-8'))
    } catch (e) {
      logger.warning('Failed to parse version file for %s %s: %s', componentName, version, String(e))
      return null
    }

    // 优先从内存缓存取回组件实例
    const cachedComponent = this._componentCache.get(`${componentName}:${version}`)
    if (cachedComponent !== undefined) {
      versionData['component'] = cachedComponent
    } else {
      const componentConfig = versionData['component_config']
      if (componentConfig && !('component' in versionData)) {
        versionData['component'] = VersionedComponentStore._deserializeComponentConfig(componentConfig)
      }
    }

    return versionData
  }

  /** 列出组件的所有版本。 */
  listVersions(componentName: string): string[] {
    return this._loadVersionsIndex(componentName)
  }

  /** 获取组件的最新版本号。 */
  getLatestVersion(componentName: string): string | null {
    const versions = this._loadVersionsIndex(componentName)
    if (versions.length === 0) {
      return null
    }
    return versions[versions.length - 1]
  }
}
